src/charm.py

```diff
@@ -59,93 +61,96 @@ def _training_operator_layer(self) -> Layer:
-        """
-        self.unit.status = MaintenanceStatus("Setting auth configuration")
+    def _create_resource(self, resource_type: str, context: dict=None) -> None:
+        """Helper method to create Kubernetes resources."""
         client = Client()
-        context = {"namespace": self._namespace, "app_name": self._name}
-
-        with open(Path(self._src_dir) / "auth_manifests.yaml") as f:
+        with open(Path(self._src_dir) / self._resource_files[resource_type]) as f:
             for obj in codecs.load_all_yaml(f, context=context):
-                try:
-                    client.create(obj)
-                except ApiError as e:
-                    logging.error(
-                        f"Creating {obj.metadata.name} failed: {e.status.reason}"
-                    )
-                    raise
-        logging.info("Auth resources successfully created.")
+                client.create(obj)
+
+    def _patch_resource(self, resource_type: str, context: dict=None) -> None:
+        """Helper method to patch Kubernetes resources."""
+        client = Client()
+        with open(Path(self._src_dir) / self._resource_files[resource_type]) as f:
+           for obj in codecs.load_all_yaml(f, context=context):
+               client.patch(type(obj), obj.metadata.name, obj, patch_type=PatchType.MERGE)
 
     def _create_crds(self) -> None:
         """Creates training-jobs CRDs.
 
         Raises:
             ApiError: if creating any of the CRDs fails.
         """
-        self.unit.status = MaintenanceStatus("Creating CRDs")
-        client = Client()
-
-        with open(Path(self._src_dir) / "crds_manifests.yaml") as f:
-            for obj in codecs.load_all_yaml(f):
-                try:
-                    client.create(obj)
-                    logging.info(f"{obj.metadata.name} CRD successfully created.")
-                except ApiError as e:
-                    if e.status.reason == "AlreadyExists":
-                        logging.info(
-                            f"{obj.metadata.name} CRD already present. It will be used by the operator."
-                        )
-                    else:
-                        logging.error(
-                            f"Creating {obj.metadata.name} failed: {e.status.reason}"
-                        )
-                        raise
+        try:
+            self._create_resource(resource_type="crds")
+        except ApiError as e:
+            if e.status.reason == "AlreadyExists":
+                logging.info(
+                    f"{e.status.details.name} CRD already present. It will be used by the operator."
+                )
+            else:
+                raise
 
     def _on_install(self, _):
-        """Event handler for on-install events."""
+        """Event handler for InstallEvent."""
+        if not self._container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for pod startup to complete")
```

Comment:
If we return from the install hook, will we ever come back here and install things again?  Is this a case where we should also `defer`?  I'm fuzzy on what should be done if install is called when we aren't ready.  @DomFleischmann @knkski ?